File: Bridge_Logger.cs
// The Bridge design pattern is used to separate the implementation from the
// abstract interface presented to the caller.  This is also colloquially
// known as the "pimpl" pattern, since the abstract interface forwards all
// method calls to an underlying implementation via a "pointer to the
// implementation" or "pimpl".
//
// The Bridge pattern separates the abstraction from the implementation so
// that both abstraction and, more importantly, implementation can vary
// without affecting the other side of the bridge.  The Bridge pattern is not
// limited to a single class with one or more implementations.  The abstract
// side can have a hierarchy of classes inheriting from a base abstraction.
// The implementation side can also have a parallel hierarchy of inherited
// classes.
//
// The key idea is the caller is presented with an abstract class that knows
// how to create a specific implementation based on input from the caller.
// The definition of the implementation is completely hidden from the caller,
// who sees only the abstract class.
//
// That abstract class must be a class since it must contain some kind of
// reference to the underlying implementation.
//
// In this example, the Logger class is what the program uses for all logging.
// The program logs information at various levels of granularity to some
// output.  The program doesn't care where the logged output ends up.
//
// The Logger abstract class can make use of three different implementations
// of loggers: File, Console, Null.  Each of these loggers implements the
// ILogger interface so the Logger object doesn't have to care about the
// implementation details.  And the user of the Logger class has no idea of
// how the Logger class does what it does, only that it provides the same
// methods regardless of the actual implementation.
//
// In this way, the Logger class is the bridge between the program and the
// logger implementations.
//
// Note: The Bridge_NullLogger is an example of the Null Object pattern.
//
// An alternative implementation would use a class factory method to take
// the LoggerType and optional parameter as arguments and return an actual
// implementation represented by the ILogger class.  This eliminates the need
// for a bridge class altogether and completely hides any details of the
// implementation.
//
// Basically, the use of an interface and a class factory can hide any
// implementation and it works in any programming language that supports the
// idea of a base class or interface.

using System;

namespace DesignPatternExamples
{
    /// <summary>
    /// Represents the logger object to be used in the program.
    /// </summary>
    /// <remarks>This class wraps different implementations of loggers to show
    /// the Bridge pattern.</remarks>
    internal class Logger : IDisposable
    {
        /// <summary>
        /// A value passed to LoggerClassFactory.CreateLogger() to specify the type of
        /// logger to create.
        /// </summary>
        public enum LoggerTypes
        {
            /// <summary>
            /// Log to nowhere, that is, throw out all logging.  No additional parameters.
            /// </summary>
            ToNull,

            /// <summary>
            /// Log to a file.  One additional parameter: the name of the file to log to.
            /// </summary>
            ToFile,

            /// <summary>
            /// Log to the console.  No additional parameters.
            /// </summary>
            ToConsole
        }

        /// <summary>
        /// Constructor that takes a LoggerTypes value and optional arguments to
        /// create a new Logger class.
        /// </summary>
        /// <param name="loggerType">A value from the LoggerTypes enum specifying the
        /// type of logger object to create.</param>
        /// <param name="arguments">Zero or more additional arguments that some logger types require.
        /// For example, a file logger requires a filename.</param>
        public Logger(LoggerTypes loggerType, params string[] arguments)
        {
            _logger = null;
            switch (loggerType)
            {
                case LoggerTypes.ToNull:
                    _logger = NullLogger.CreateNullLogger();
                    break;

                case LoggerTypes.ToConsole:
                    _logger = ConsoleLogger.CreateConsoleLogger();
                    break;

                case LoggerTypes.ToFile:
                    {
                        if (arguments != null && arguments.Length >= 1)
                        {
                            _logger = FileLogger.CreateFileLogger(arguments[0]);
                        }
                        else
                        {
                            string msg = String.Format("A filename must be specified for the {0} logger type.", loggerType);
                            throw new ArgumentException(msg, "arguments");
                        }
                    }
                    break;

                default:
                    {
                        string msg = String.Format("The logger type '{0}' is not recognized.  Cannot construct a Logger.", loggerType);
                        throw new ArgumentException(msg, "loggerType");
                    }
            }
        }


        /// <summary>
        /// Log trace messages to the configured output.
        /// </summary>
        /// <param name="message">The message to log.</param>
        public void LogTrace(string message)
        {
            if (_logger != null)
            {
                _logger.LogTrace(message);
            }
        }

        /// <summary>
        /// Log informational messages to the configured output.
        /// </summary>
        /// <param name="message">The message to log.</param>
        public void LogInfo(string message)
        {
            if (_logger != null)
            {
                _logger.LogInfo(message);
            }
        }

        /// <summary>
        /// Log error messages to the configured output.
        /// </summary>
        /// <param name="message">The message to log.</param>
        public void LogError(string message)
        {
            if (_logger != null)
            {
                _logger.LogError(message);
            }
        }

        #region IDisposable Members

        /// <summary>
        /// Dispose of this logger object in a friendly way.
        /// </summary>
        void IDisposable.Dispose()
        {
            if (_logger != null)
            {
                if (_logger is IDisposable)
                {
                    ((IDisposable)_logger).Dispose();
                }
            }
        }

        #endregion

        private ILogger _logger;
    }

}

File: python/DesignPatternExamples_python/bridge/bridge_logger.py
##  @file
#  @brief
#  The @ref DesignPatternExamples_python.bridge.bridge_logger.Logger "Logger"
#  bridge class as used in the @ref bridge_pattern.

from enum import Enum
from .bridge_loggerinterface import ILogger
from .bridge_filelogger import FileLogger
from .bridge_consolelogger import ConsoleLogger
from .bridge_nulllogger import NullLogger

##  Represents the bridge logger object to be used in the program.
#
#  This class wraps different implementations of loggers to show
#  the @ref bridge_pattern.
class Logger(ILogger):

    ##  A value passed to LoggerClassFactory.CreateLogger() to specify the type of
    #  logger to create.
    class LoggerTypes(Enum):
        ##  Log to nowhere, that is, throw out all logging.  No additional parameters.
        ToNull = 0,

        ##  Log to a file.  One additional parameter: the name of the file to log to.
        ToFile = 1,

        ##  Log to the console.  No additional parameters.
        ToConsole = 2


    ## Constructor that takes a LoggerTypes value and optional arguments to
    #  create a new Logger class.
    #
    #  @param loggerType
    #         A value from the LoggerTypes enum specifying the type of logger
    #         object to create.
    #  @param arguments
    #         Zero or more additional arguments that some logger types require.
    #         For example, a file logger requires a filename.
    def __init__(self, loggerType : LoggerTypes, *arguments):
        self._logger = None
        match (loggerType):
            case Logger.LoggerTypes.ToNull:
                self._logger = NullLogger.CreateLogger()

            case Logger.LoggerTypes.ToConsole:
                self._logger = ConsoleLogger.CreateLogger()

            case Logger.LoggerTypes.ToFile:
                if arguments:
                    self._logger = FileLogger.CreateLogger(arguments[0])
                else:
                    msg = "A filename must be specified for the {0} logger type.".format(loggerType)
                    raise ValueError(msg)

            case _: # default
                msg = "The logger type '{0}' is not recognized.  Cannot construct a Logger.".format(loggerType)
                raise ValueError(msg)

    ## @var _logger
    #       Instance of the desired logger type

    ## Entry function used in the `with` statement to initialize an instance
    #  of the reader/writer.
    def __enter__(self):
        return self


    ## Exit function automatically called when used in the `with` statement.
    def __exit__(self, *args):
        if self._logger:
            self._logger.Close()
            self._logger = None

    #-------------------------------------------------------------------------
    # ILogger interface implementation
    #-------------------------------------------------------------------------

    ## Close the logger and release resources
    #
    #  Call only if the Logger was not created in a `with` statement.
    def Close(self):
        if self._logger:
            self._logger.Close()
            self._logger = None

    ##  Log trace messages to the configured output.
    #
    #  @param message
    #         The message to log.
    def LogTrace(self, message : str):
        if self._logger:
            self._logger.LogTrace(message)



    ##  Log informational messages to the configured output.
    #
    #  @param message
    #         The message to log.
    def LogInfo(self, message : str):
        if self._logger:
            self._logger.LogInfo(message)


    ##  Log error messages to the configured output.
    #
    #  @param message
    #         The message to log.
    def LogError(self, message : str):
        if self._logger:
            self._logger.LogError(message)


